### PR TITLE
Anchor commitments, new payment basepoint

### PR DESCRIPTION
**Pull request by bitromortac**
This PR implements a newer commitment transaction type [1]. Anchor commitments are commitment transactions that attach a dusty output per realized `to_local`/`to_remote` output, in order for the commitment transaction to be CPFP-bumpable in a force close scenario. Also HTLC transactions are affected, which are fee-amendable allowing for more inputs via the anyonecanpay sighash. The original proposal of anchor commitments [1] was not safe enough [2] which is why `option_anchors_zero_fee_htlc_tx` was devised [3]. Our goal is to go directly to this version, to not having to deal with legacy developments. The PR passes unit and regtest tests and is ready for initial review or feedback (see Todos).

Key changes:
- The `Channel` object now carries a `has_anchors` property which is used to decide on scripts and locktimes, as well as sweeping behavoir.
- Adapt commitment transactions for anchors (anchors, to_remote script 1-CSV, fee estimation, output values) (03-transactions)
- Adapt HTLC transactions (nsequence numbers, 1 CSV-lock, `sighash.ANYONECANPAY+sighash.SINGLE`) (03-transactions)
- New anchor channels have a `payment_basepoint` derived from the funding pubkey and a static secret to be sweepable in the event of a force close due to (onchain) backups

Testing:
- Reproduce signatures from RFC test vectors for commitment and HTLC transactions (requires different sighash flags) (03-transactions)
- Manually tested against lnd v0.11 (last active version of original anchors) - open/close/force close/payment
- New integration test framework to run Electrum peer components inside a single interpreter (better debugging and easier introspection), with additional detailed channel usage scenarios. Tests are not running in sequence right now and are still subject to changes. Some tests are redundant and could be removed.

Other:
- The watchtower is not updated to sweep to_remote as this is not time critical
- Backups don't need an upgrade as we can see at sweep time, whether we deal with a ctx that has anchors

Todo:
- [x] Crosscheck RFC
- [x] Implement zero fee HTLCs and do fee management at sweep time
- [ ] Reserve some UTXOs for fee bumping
- [x] Compatibility testing with zero-fee-htlc (once zero-fee is implemented)
- [x] Wallet upgrading mechanics, compatibilty wallets with old/new channels
- [x] Keep tests for both commitment versions (static_remotekey/anchors)?
- [x] Make commitment type configurable?
- [x] Determine fee for HTLC-transaction bumping
- Sweep the anchor outputs (05-onchain) (omitted for this PR, because we still rely on the `update_fee` message)
- Refactor to have explicit commitment and channel types [4] (another PR) 

[1] Anchors PR: https://github.com/lightningnetwork/lightning-rfc/pull/688
[2] https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-September/002796.html
[3] Zero-fee HLTCs: https://github.com/lightningnetwork/lightning-rfc/pull/824
[4] Channel types: https://github.com/lightningnetwork/lightning-rfc/pull/880

### CODE REVIEW

File: electrum/lnsweep.py [hunk 1]
```diff
@@ -328,35 +337,105 @@ def analyze_ctx(chan: 'Channel', ctx: Transaction):
...
+        pubkey2 = script[2][1]
+        return [pubkey1, pubkey2]
+
+
+def create_sweeptx_their_backup_ctx(
+            *, chan: 'ChannelBackup',
+            ctx: Transaction,
+            sweep_address: str) -> Optional[Dict[str, SweepInfo]]:
+    txs = {}  # type: Dict[str, SweepInfo]
+    """Sweep to_remote, if we use anchors."""
```
Comment (SomberNight) [inline electrum/lnsweep.py:357]: Maybe elaborate a bit on the scenario in the docstring.
e.g. 
> If we only have a backup, and the remote force-closed with their ctx, and anchors are enabled, we need to sweep to_remote.

File: electrum/lnchannel.py [hunk 2]
```diff
@@ -525,6 +546,9 @@ def is_static_remotekey_enabled(self) -> bool:
         # their local config is not static)
         return False
 
+    def has_anchors(self) -> Optional[bool]:
+        return None
```
Comment (ecdsa) [inline electrum/lnchannel.py:550]: is that method used? 
in `create_sweeptx_their_backup_ctx` you are using `lnutil.ctx_has_anchors()` instead

**Review by SomberNight — COMMENTED** [review]
So many changes :P but looks good overall.
See comments inline.

File: electrum/lnchannel.py (hunk 2 above)
Comment (bitromortac) [reply electrum/lnchannel.py:550]: It's not being used, but it's an `abstractmethod`, so has to be implemented. With the `None` return I wanted to indicate that we don't know about the channel having anchors.

File: electrum/lnchannel.py [hunk 3]
```diff
@@ -322,9 +324,14 @@ def update_closed_state(self, *, funding_txid: str, funding_height: TxMinedInfo,
     def sweep_address(self) -> str:
         # TODO: in case of unilateral close with pending HTLCs, this address will be reused
         addr = None
-        if self.is_static_remotekey_enabled():
+        if self.has_anchors():
+            try:
+                addr = self.lnworker.wallet.get_new_sweep_address_for_channel()
+            except AttributeError:  # fails in tests with MockWallet
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/lnchannel.py]: Done

File: electrum/tests/test_lnutil.py [hunk 4]
```diff
@@ -38,6 +43,285 @@
 # funding wscript = 5221023da092f6980e58d2c037173180e9a465476026ee50f96695963e8efe436f54eb21030e9f7b623d2ccc7c9bd44d66d5ce21ce504c0acf6385a132cec6d3c39fa711c152ae
 
 
+# https://github.com/lightningnetwork/lightning-rfc/blob/master/03-transactions.md#appendix-f-commitment-and-htlc-transaction-test-vectors-anchors
+ANCHOR_TEST_VECTORS = [
+    {
+        "Name": "simple commitment tx with no HTLCs",
+        "LocalBalance": 7000000000,
+        "RemoteBalance": 3000000000,
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/tests/test_lnutil.py]: Done

File: electrum/lnpeer.py [hunk 5]
```diff
@@ -502,6 +502,10 @@ def is_static_remotekey(self):
     def is_upfront_shutdown_script(self):
         return self.features.supports(LnFeatures.OPTION_UPFRONT_SHUTDOWN_SCRIPT_OPT)
 
+    def use_anchors(self) -> bool:
+        return self.features.supports(LnFeatures.OPTION_ANCHOR_OUTPUTS_OPT) \
+               and self.lnworker.features.supports(LnFeatures.OPTION_ANCHOR_OUTPUTS_OPT)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/lnpeer.py:507]: Yes, here I wanted to ask how we want to transition to anchors? The second condition was also a reminder for myself.

File: electrum/tests/test_lnutil.py [hunk 6]
```diff
@@ -38,6 +43,285 @@
 # funding wscript = 5221023da092f6980e58d2c037173180e9a465476026ee50f96695963e8efe436f54eb21030e9f7b623d2ccc7c9bd44d66d5ce21ce504c0acf6385a132cec6d3c39fa711c152ae
 
 
+# https://github.com/lightningnetwork/lightning-rfc/blob/master/03-transactions.md#appendix-f-commitment-and-htlc-transaction-test-vectors-anchors
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/tests/test_lnutil.py]: Agree, changed

File: electrum/lnutil.py [hunk 7]
```diff
@@ -962,6 +1131,18 @@ class LnFeatures(IntFlag):
...
+    OPTION_ANCHOR_OUTPUTS_OPT = 1 << 21
+    _ln_feature_direct_dependencies[OPTION_ANCHOR_OUTPUTS_OPT] = {OPTION_STATIC_REMOTEKEY_OPT}
+    _ln_feature_contexts[OPTION_ANCHOR_OUTPUTS_REQ] = (LNFC.INIT | LNFC.NODE_ANN)
+    _ln_feature_contexts[OPTION_ANCHOR_OUTPUTS_OPT] = (LNFC.INIT | LNFC.NODE_ANN)
+
+    OPTION_ANCHORS_ZERO_FEE_HTLC_REQ = 1 << 22
+    OPTION_ANCHORS_ZERO_FEE_HTLC_OPT = 1 << 23
+    _ln_feature_direct_dependencies[OPTION_ANCHORS_ZERO_FEE_HTLC_OPT] = {OPTION_STATIC_REMOTEKEY_OPT}
+    _ln_feature_contexts[OPTION_ANCHORS_ZERO_FEE_HTLC_REQ] = (LNFC.INIT | LNFC.NODE_ANN)
+    _ln_feature_contexts[OPTION_ANCHORS_ZERO_FEE_HTLC_OPT] = (LNFC.INIT | LNFC.NODE_ANN)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/lnutil.py:1146]: Didn't spot that, thanks

File: electrum/lnsweep.py [hunk 8]
```diff
@@ -328,35 +337,105 @@ def analyze_ctx(chan: 'Channel', ctx: Transaction):
     return ctn, their_pcp, is_revocation, per_commitment_secret
 
 
+def extract_funding_pubkeys_from_ctx(txin: TxInput) -> Optional[List[ecc.ECPubkey]]:
+    witness = txin.witness_elements()
+    if len(witness) == 4:
+        # we have OP_2(82) pk1 pk2 OP_2 OP_CHECKMULTISIG(174)
+        # TODO: could match_script_against_template
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/lnsweep.py]: Ok, have done it.

File: electrum/lnsweep.py [hunk 9]
```diff
@@ -328,35 +337,105 @@ def analyze_ctx(chan: 'Channel', ctx: Transaction):
     return ctn, their_pcp, is_revocation, per_commitment_secret
 
 
+def extract_funding_pubkeys_from_ctx(txin: TxInput) -> Optional[List[ecc.ECPubkey]]:
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/lnsweep.py]: Oh yes, initially I returned `ECPubkey`s

File: electrum/lnutil.py [hunk 10]
```diff
@@ -719,41 +798,77 @@ class Direction(IntFlag):
...
+            to_remote_amt_msat -= 2 * FIXED_ANCHOR_SAT * 1000
+        # only include anchors for outputs that are not below dust limit
+        if to_local_amt_msat // 1000 >= dust_limit_sat:
+            anchor_outputs.append(PartialTxOutput(scriptpubkey=bfh(local_anchor_script), value=FIXED_ANCHOR_SAT))
+        if to_remote_amt_msat // 1000 >= dust_limit_sat:
+            anchor_outputs.append(PartialTxOutput(scriptpubkey=bfh(remote_anchor_script), value=FIXED_ANCHOR_SAT))
+
+    # if funder cannot afford feerate, their output might go negative, so take max(0, x) here
+    to_local_amt_msat = max(0, to_local_amt_msat)
+    to_remote_amt = max(0, to_remote_amt_msat)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/lnutil.py]: Done

File: electrum/lnsweep.py (hunk 1 above)
Comment (bitromortac) [reply electrum/lnsweep.py]: Done

**Comment by bitromortac** [comment]
I submitted changes via fixup commits, will squash in the end, or should I already?

File: electrum/lnpeer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SomberNight) [reply electrum/lnpeer.py:507]: Well there are three ~reasonable options I can see:
1. we require anchors_zero_fee for newly opened channels
2. we open anchors_zero_fee channels whenever possible, but if the remote does not support it, we can just option static_remotekey channels
3. we disable anchors_zero_fee by default, gate it with a config key; and change to one of the options above in a later release

I think the decision should depend on a number of things, such as
- how many/which nodes on the network will already support anchors_zero_fee,
- how confident we are that our code behaves correctly, 
- how the UX of reserving UTXOs in the wallet looks like,
- whether we can batch HTLC txs (not really for fee-minimisation but rather re number of UTXOs we need to reserve)

So atm I am not sure.

**Comment by SomberNight** [comment]
> I submitted changes via fixup commits, will squash in the end, or should I already?

Feel free to do whatever is convenient.
One remark: if you rebase on master for any reason, it's better to do that in a separate force-push of its own (instead of combining with other changes that would require a force-push). That way, it's easier to see changes in the github UI.

File: electrum/lnsweep.py [hunk 11]
```diff
@@ -328,35 +337,107 @@ def analyze_ctx(chan: 'Channel', ctx: Transaction):
     return ctn, their_pcp, is_revocation, per_commitment_secret
 
 
+def extract_funding_pubkeys_from_ctx(txin: TxInput) -> Tuple[bytes, bytes]:
+    """Extract the two funding pubkeys from the published commitment transaction.
+
+    We expect to see a witness script of: OP_2 pk1 pk2 OP_2 OP_CHECKMULTISIG"""
+    elements = txin.witness_elements()
+    witness_script = elements[-1]
+    assert match_script_against_template(witness_script, SCRIPTPUBKEY_TEMPLATE_FUNDING)
```
Comment (SomberNight) [inline electrum/lnsweep.py:346]: nit. `SCRIPTPUBKEY_TEMPLATE_FUNDING` as a name is a bit confusing:
"scriptPubKey" is conventionally the name for the output script directly visible in the chain (i.e. the script in the output, so what a bitcoin address directly maps to).

So e.g. these are indeed templates for scriptPubKeys:
https://github.com/spesmilo/electrum/blob/437a9e435847431e8654a6e02fc304a714826533/electrum/transaction.py#L436-L442
The templates here are also misnamed IMHO, as they are matching witnessScripts (note difference between witness and witnessScript):
https://github.com/spesmilo/electrum/blob/437a9e435847431e8654a6e02fc304a714826533/electrum/submarine_swaps.py#L37

Maybe just call it `SCRIPT_TEMPLATE_FUNDING`?

Comment (bitromortac) [reply electrum/lnsweep.py]: Makes sense, done.

**Comment by bitromortac** [comment]
> > I submitted changes via fixup commits, will squash in the end, or should I already?
> 
> Feel free to do whatever is convenient. One remark: if you rebase on master for any reason, it's better to do that in a separate force-push of its own just for the rebase (instead of combining with other changes that would require a force-push). That way, it's easier to see changes in the github UI.

I squashed the fixups and rebased on master separately.

There's a new commit, which switches to the `anchors-zero-htlc-fee` type. I have tested this together with the latest LND master. The only issue for now is how to deal with password protected wallets, where we are not able to fee-bump HTLCs automatically (or bump commitment transactions once we phase out `update_fee`).

**Comment by bitromortac** [comment]
I reorganized the commits a bit and think this is ready for a second pass. Tests for anchor outputs are disabled right now (can be controlled via `TEST_ANCHOR_CHANNELS`).

File: electrum/lnsweep.py [hunk 12]
```diff
@@ -461,19 +567,97 @@ def create_htlctx_that_spends_from_our_ctx(
...
+            outputs.insert(0, htlc_txout)
+        final_htlc_tx = PartialTransaction.from_io(
+            inputs,
+            outputs,
+            locktime=maybe_zero_fee_htlc_tx.locktime,
+            version=maybe_zero_fee_htlc_tx.version,
+            BIP69_sort=False
+        )
+
+        # sign the fee input(s): TODO: how to deal with password-protected wallets?
```
Comment (SomberNight) [inline electrum/lnsweep.py:631]: I see this is not using `wallet.wallet_password` yet

File: electrum/gui/qt/installwizard.py [hunk 13]
```diff
@@ -336,7 +337,7 @@ def run_user_interaction_loop():
                                               msg=_('Failed to decrypt using this hardware device.') + '\n' +
                                                   _('If you use a passphrase, make sure it is correct.'))
                             self.reset_stack()
-                            return self.select_storage(path, get_wallet_from_daemon)
+                            return
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/gui/qt/installwizard.py]: Restored.

File: electrum/lnworker.py [hunk 14]
```diff
@@ -648,6 +648,20 @@ def __init__(self, wallet: 'Abstract_Wallet', xprv):
             self.set_invoice_status(payment_hash.hex(), PR_INFLIGHT)
 
         self.trampoline_forwarding_failures = {} # todo: should be persisted
+        self.wallet_password = None  # used for automatic signing in case of anchor channels
+
+    def maybe_enable_anchors_store_password(self, password):
+        # TODO: make this a per-wallet config option?
+        # for anchor commitments we need the password to be able to spend wallet UTXOs
+        if self.config.get('enable_anchor_channels'):
+            if self.wallet.is_watching_only():
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/lnworker.py]: Have added your suggestions, thank you.

File: electrum/tests/test_lnchannel.py [hunk 15]
```diff
@@ -33,22 +33,23 @@
...
-from electrum.lnutil import SENT, LOCAL, REMOTE, RECEIVED
+from electrum.lnutil import SENT, LOCAL, REMOTE, RECEIVED, effective_htlc_tx_weight
 from electrum.logging import console_stderr_handler
 from electrum.lnchannel import ChannelState
 from electrum.json_db import StoredDict
 from electrum.coinchooser import PRNG
 
 from . import ElectrumTestCase
 
+TEST_ANCHOR_CHANNELS = False
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/tests/test_lnchannel.py]: Tests that concern anchor commitment transactions (`test_lnchannel.py`, `test_lnpeer.py`, and `regtest.py`) are now duplicated.

File: electrum/tests/regtest/regtest.sh [hunk 16]
```diff
@@ -2,6 +2,7 @@
 export HOME=~
 set -eu
 
+TEST_ANCHOR_CHANNELS=False
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/tests/regtest/regtest.sh]: Done, I'm setting an environment variable in `subprocess.Popen`. Not sure if updating of the environment dict is the best approach.

File: electrum/tests/integration.py [hunk 17]
```diff
@@ -0,0 +1,555 @@
+import asyncio
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/tests/integration.py:1]: The tests could be placed into regtest.py and should be executable in the same way (with bitcoind and electrumx requirements). I'm still thinking about which tests to keep. I still have to remove some sleeps and test more for specific outcomes.

File: electrum/lnsweep.py (hunk 12 above)
Comment (bitromortac) [reply electrum/lnsweep.py]: Somehow got lost during rebasing :+1:

**Comment by bitromortac** [comment]
After the recent update all tests passed, but I had to run the watchtower test twice (needs further investigation). I'm still working on the HTLC issue we discussed.

**Comment by bitromortac** [comment]
> After the recent update all tests passed, but I had to run the watchtower test twice (needs further investigation). I'm still working on the HTLC issue we discussed.

I tried to add all the edge cases concerning HTLC transaction batching. Do we need an explicit test for this? I also added a test for the watchtower in case of anchor outputs with an HTLC in the ctx (see `integration.TestLightningABC.test_watchtower_htlc`)

File: electrum/tests/integration.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (bitromortac) [reply electrum/tests/integration.py]: I have opened another PR for the tests.

**Comment by ecdsa** [comment]
Note: I'm going to rebase this in a new branch. Since it is not possible to change the source branch of a PR, I guess this means that the conversation will have to be continued in a new PR

**Comment by ecdsa** [comment]
new branch, rebased on current master: https://github.com/spesmilo/electrum/commits/anchor_commitments_2022

**Comment by ecdsa** [comment]
new rebase: https://github.com/spesmilo/electrum/commits/anchor_commitments_2024

**Comment by SomberNight** [comment]
@ecdsa Could you make a separate PR for your rebase? I cannot usefully add comments to it without that.
(note to self: lnchannel.sweep_address looks broken in the rebase)

**Comment by ecdsa** [comment]
opened PR  #9264

File: electrum/lnchannel.py [hunk 18]
```diff
@@ -1193,25 +1193,42 @@ def consider_ctx(*, ctx_owner: HTLCOwner, is_htlc_dust: bool) -> int:
...
-                max_send_msat = sender_balance_msat - sender_reserve_msat - ctx_fees_msat[sender]
+                    has_anchors=self.has_anchors())[sender]
+                max_send_msat -= fee_spike_buffer
+            # we can't enforce the fee spike buffer on the remote party
+            elif sender == initiator == REMOTE:
+                max_send_msat -= ctx_fees_msat[sender]
+
+            # initiator pays for anchor outputs
+            if sender == initiator and self.has_anchors():
+                max_send_msat -= 2 * FIXED_ANCHOR_SAT
```
Comment (ecdsa) [inline electrum/lnchannel.py:1223]: note: needs to be multiplied by 1000

**Comment by ecdsa** [comment]
commits were reworked and merged in #9264
